examples/jsm/materials/MeshGouraudMaterial.js

```diff
@@ -0,0 +1,303 @@
+			vec3 totalEmissiveRadiance = emissive;
+
+			#include <logdepthbuf_fragment>
+			#include <map_fragment>
+			#include <color_fragment>
+			#include <alphamap_fragment>
+			#include <alphatest_fragment>
+			#include <specularmap_fragment>
+			#include <emissivemap_fragment>
+
+			// accumulation
+
+			#ifdef DOUBLE_SIDED
+
+				reflectedLight.indirectDiffuse += ( gl_FrontFacing ) ? vIndirectFront : vIndirectBack;
+
+			#else
+
+				reflectedLight.indirectDiffuse += vIndirectFront;
+
+			#endif
+
+			#include <lightmap_fragment>
+
+			reflectedLight.indirectDiffuse *= BRDF_Lambert( diffuseColor.rgb );
+
+			#ifdef DOUBLE_SIDED
+
+				reflectedLight.directDiffuse = ( gl_FrontFacing ) ? vLightFront : vLightBack;
+
+			#else
+
+				reflectedLight.directDiffuse = vLightFront;
+
+			#endif
+
+			reflectedLight.directDiffuse *= BRDF_Lambert( diffuseColor.rgb ) * getShadowMask();
+
+			// modulation
+
+			#include <aomap_fragment>
+
+			vec3 outgoingLight = reflectedLight.directDiffuse + reflectedLight.indirectDiffuse + totalEmissiveRadiance;
+
+			#include <envmap_fragment>
+
+			#include <output_fragment>
+			#include <tonemapping_fragment>
+			#include <encodings_fragment>
+			#include <fog_fragment>
+			#include <premultiplied_alpha_fragment>
+			#include <dithering_fragment>
+
+		}`
+
+};
+
+//
+
+class MeshGouraudMaterial extends ShaderMaterial {
+
+	constructor( parameters ) {
+
+		super();
+
+		this.isMeshGouraudMaterial = true;
+
+		this.type = 'MeshGouraudMaterial';
+
+		//this.color = new THREE.Color( 0xffffff ); // diffuse
+
+		//this.map = null;
+
+		//this.lightMap = null;
+		//this.lightMapIntensity = 1.0;
+
+		//this.aoMap = null;
+		//this.aoMapIntensity = 1.0;
+
+		//this.emissive = new THREE.Color( 0x000000 );
+		//this.emissiveIntensity = 1.0;
+		//this.emissiveMap = null;
+
+		//this.specularMap = null;
+
+		//this.alphaMap = null;
+
+		//this.envMap = null;
+		this.combine = MultiplyOperation; // combine has no uniform
+		//this.reflectivity = 1;
+		//this.refractionRatio = 0.98;
+
+		//this.flipEnvMap = - 1; // if envMap is a CubeTexture and not a RenderTargetTexture, set to - 1, otherwise set to 1
+
+		this.fog = false; // set to use scene fog
+		this.lights = true; // set to use scene lights
+		this.clipping = false; // set to use user-defined clipping planes
+
+		const shader = GouraudShader;
+
+		this.defines = Object.assign( {}, shader.defines );
+		this.uniforms = UniformsUtils.clone( shader.uniforms );
+		this.vertexShader = shader.vertexShader;
+		this.fragmentShader = shader.fragmentShader;
+
+		const exposePropertyNames = [
+			'map', 'lightMap', 'lightMapIntensity', 'aoMap', 'aoMapIntensity',
+			'emissive', 'emissiveIntensity', 'emissiveMap', 'specularMap', 'alphaMap',
+			'envMap', 'reflectivity', 'refractionRatio', 'opacity', 'diffuse', 'flipEnvMap'
+		];
+
+		for ( const propertyName of exposePropertyNames ) {
+
+			Object.defineProperty( this, propertyName, {
```

Comment:
Wouldn't this be too slow?

Reply:
The pattern is used in `MMDToonMaterial`, so I do not think it is an issue.

Reply:
I can't see a reason this would be a bottleneck in practice. We are relying on getters/setters in many places. If this turns out to be a performance problem that affects users, we can make adjustments. If anyone would like to do benchmarks with/without this part that's always welcome, but it does not seem likely enough to change speculatively.

Reply:
The problem is that Object.defineProperty is slow at construction time, not that getters/setters are slow at access time.

Reply:
I think @LeviPesin is referring to #21284, however I think in this context `Object.defineProperty` is acceptable since the user is not going to create many MeshGouraudMaterials sequentially.